##### Grupo
* Danilo F Duarte - 203085
* Darci Belon - 203022
* Eduardo Mourão - 203072
* Fábio Henrique - 203021
* Victor Ciccolani - 203094

In [151]:
from sklearn.ensemble import BaggingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from sklearn import datasets
import matplotlib.pyplot
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.tree import DecisionTreeClassifier
import time
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif
from statsmodels.stats.contingency_tables import mcnemar
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings(action='ignore', category=DeprecationWarning)

In [126]:
#import da base
digits = datasets.load_digits()
X = digits.data
y = digits.target

print(X.shape)
print(y.shape)

(1797, 64)
(1797,)


### Declarando funções que serão usadas

In [135]:
#Declarando o conteúdo da atividade em funções (KNN, decision tree e random forest)
tipo = []
dados = []
acc =[]
time_l = []


def knn_bagging(Xtrain,ytrain,Xtest,ytest, k):
    #dados sem reducao
    start = time.time()
    model = KNeighborsClassifier()
    scores = cross_val_score(model, Xtrain, ytrain, cv=5)
    print('Acurácia de KNeighbors simples:', scores.mean())
    end = time.time()
    print('Tempo:',end - start)
    model.fit(Xtrain, ytrain)
    y_pred_1 = model.predict(Xtest)
    
    tip1 = 'Acurácia de KNeighbors simples:'
    a1 = scores.mean()
    t1 = end - start
    acc.append(a1)
    time_l.append(t1)
    tipo.append(tip1)
    d1 = k
    dados.append(d1)
    
    start = time.time()
    model = BaggingClassifier(KNeighborsClassifier(), max_samples=0.5, max_features=0.5, random_state = 42)
    scores = cross_val_score(model, Xtrain, ytrain, cv=5)
    print('Acurácia de KNeighbors Bagging (c/ 10 estimators):', scores.mean())
    end = time.time()
    print('Tempo:',end - start)
    model.fit(Xtrain, ytrain)
    y_pred_2 = model.predict(Xtest)
    
    tip2 = 'Acurácia de KNeighbors Bagging (c/ 10 estimators):'
    a2 = scores.mean()
    t2 = end - start
    acc.append(a2)
    time_l.append(t2)
    tipo.append(tip2)
    d2 = k
    dados.append(d2)
    
    
    start = time.time()
    model = BaggingClassifier(KNeighborsClassifier(), max_samples=0.5, max_features=0.5, n_estimators=100, random_state = 42)
    scores = cross_val_score(model, Xtrain, ytrain, cv=5)
    print('Acurácia de KNeighbors Bagging (c/ 100 estimators):', scores.mean())
    end = time.time()
    print('Tempo:',end - start)
    model.fit(Xtrain, ytrain)
    y_pred_3 = model.predict(Xtest)
    
    tip3 ='Acurácia de KNeighbors Bagging (c/ 100 estimators):'
    a3 = scores.mean()
    t3 = end - start
    acc.append(a3)
    time_l.append(t3) 
    tipo.append(tip3)
    d3 = k
    dados.append(d3)
    
    print('-------------------------------------')
    print('Teste do McNemar')
    testaMcNemar(ytest,y_pred_1,y_pred_2)
    testaMcNemar(ytest,y_pred_1,y_pred_3)
    testaMcNemar(ytest,y_pred_2,y_pred_3)
    
    
    return y_pred_1,y_pred_2,y_pred_3
#################################################################################################
    
def decision_tree(Xtrain,ytrain,Xtest,ytest, k):
    start = time.time()
    model = DecisionTreeClassifier(max_depth=None, min_samples_split=2, random_state=0)
    scores = cross_val_score(model, Xtrain, ytrain, cv=5)
    print('Acurácia de Decision Tree puro:', scores.mean())
    end = time.time()
    print('Tempo:',end - start)
    model.fit(Xtrain, ytrain)
    y_pred_1 = model.predict(Xtest)
    
    tip4 = 'Acurácia de Decision Tree puro:'
    a4 = scores.mean()
    t4 = end - start
    acc.append(a4)
    time_l.append(t4) 
    tipo.append(tip4)
    d4 = k
    dados.append(d4)
    
    start = time.time()
    model = RandomForestClassifier(n_estimators=50, max_depth=None, min_samples_split=2, random_state=0)
    scores = cross_val_score(model, Xtrain, ytrain, cv=5)
    print('Acurácia de Random Forest:', scores.mean())
    end = time.time()
    print('Tempo:',end - start)
    model.fit(Xtrain, ytrain)
    y_pred_2 = model.predict(Xtest)
    
    tip5 = 'Acurácia de Random Forest:'
    a5 = scores.mean()
    t5 = end - start
    acc.append(a5)
    time_l.append(t5) 
    tipo.append(tip5)
    d5 = k
    dados.append(d5)
    
    start = time.time()
    model = ExtraTreesClassifier(n_estimators=50, max_depth=None, min_samples_split=2, random_state=0)
    scores = cross_val_score(model, Xtrain, ytrain, cv=5)
    print('Acurácia de Extreme Randomized Trees:', scores.mean())
    end = time.time()
    print('Tempo:',end - start)
    model.fit(Xtrain, ytrain)
    y_pred_3 = model.predict(Xtest)
    
    tip6 = 'Acurácia de Random Forest:'
    a6 = scores.mean()
    t6 = end - start
    acc.append(a6)
    time_l.append(t6) 
    tipo.append(tip6)
    d6 = k
    dados.append(d6)
        
    print('-------------------------------------')
    print('Teste do McNemar')
    testaMcNemar(ytest,y_pred_1,y_pred_2)
    testaMcNemar(ytest,y_pred_1,y_pred_3)
    testaMcNemar(ytest,y_pred_2,y_pred_3)
    
    return y_pred_1,y_pred_2,y_pred_3
########################################################################################################

def multiples_vote(Xtrain,ytrain,Xtest,ytest):
    start = time.time()
    #Obs aumentei a max_inter da regressão logistica para tirar os warnings (deixou bem lento, tem a ver com a escala dos dados)
    clf1 = LogisticRegression(solver='lbfgs', multi_class='multinomial',max_iter = 10000, random_state=1)
    clf2 = RandomForestClassifier(n_estimators=10, random_state=1)
    clf3 = GaussianNB()

    eclf = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2), ('gnb',clf3)], voting='hard')

    for clf, label in zip([clf1, clf2, clf3, eclf], ['Logistic Regression', 'Random Forest', 'naive Bayes', 'Ensemble']):
        scores = cross_val_score(clf, X_train, y_train, cv=5)
        print("Acurácia: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))

    print('-'*20)

    eclf = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2)], voting='hard')

    for clf, label in zip([clf1, clf2, eclf], ['Logistic Regression', 'Random Forest', 'Ensemble']):
        scores = cross_val_score(clf, X_train, y_train, cv=5)
        print("Acurácia: %0.2f (+/- %0.2f) [%s]" % (scores.mean(), scores.std(), label))
    end = time.time()
    print('Tempo:',end - start)

#######################################################################################################

def gradient_boost(Xtrain,ytrain,Xtest,ytest, k):
    
    start = time.time()
    model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.3, max_depth=2, random_state=0)
    scores = cross_val_score(model, Xtrain, ytrain, cv=5)

    print('Acurácia de Gradient Boosting Tree:', scores.mean())
    end = time.time()
    print('Tempo:',end - start)
    model.fit(Xtrain, ytrain)
    y_pred_1 = model.predict(Xtest)
    
    tip7 = 'Acurácia de Gradient Boosting Tree:'
    a7 = scores.mean()
    t7 = end - start
    acc.append(a7)
    time_l.append(t7) 
    tipo.append(tip7)
    d7 = k
    dados.append(d7)
    

    return y_pred_1
    

In [136]:
#Fazer a parte do mc nemmar: A ideia é fazer 1 comparação com cada tipo (originais,redução e seleção):

#essa primeira função cria a contigence table a partir do y_test,y_pred1 e y_pred2 (resultados que vc quer comparar)
def build_contingence_table(Y, Y_pred_1, Y_pred_2):
    y1_and_y2 = 0
    y1_and_not_y2 = 0
    y2_and_not_y1 = 0
    not_y1_and_not_y2 = 0
    for y, y1, y2 in zip(Y, Y_pred_1, Y_pred_2):
        if y == y1 == y2:
            y1_and_y2 += 1
        elif y != y1 and y != y2:
            not_y1_and_not_y2 += 1
        elif y == y1 and y != y2:
            y1_and_not_y2 += 1
        elif y != y1 and y == y2:
            y2_and_not_y1 += 1
            
    contingency_table = [[y1_and_y2, y1_and_not_y2], 
                         [y2_and_not_y1, not_y1_and_not_y2]]
    
    return contingency_table
    

In [137]:
def testaMcNemar(test, pred1, pred2):
    contingence_table = build_contingence_table(test, pred1, pred2)

    import pprint

    pprint.pprint(contingence_table)

    result = mcnemar(contingence_table, exact=True)


    if result.pvalue >= 0.001:
        print('statistic=%.3f, p-value=%.3f' % (result.statistic, result.pvalue))
    else:
        print('statistic=%.3f, p-value=%.3e' % (result.statistic, result.pvalue))

    # interpretando o p-value
    alpha = 0.05
    if result.pvalue > alpha:
        print('Mesma proporção de erros (falhou em rejeitar H0)')
    else:
        print('Proporções de erros diferentes (rejeitou H0)')

### Dividindo em treino e teste para cada tipo (original, redução e seleção)

In [138]:
# Aplicando Train test split nos dados originais
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

print('Número original de atributos:', X.shape[1])

#----------------------------------------------------------------------------------------------------------

# Aplicando Train test split nos dados com reduçao de atributos (PCA)

#realizando a redução
pca = PCA(n_components=0.90, whiten=True)

X_pca = pca.fit_transform(X)

X_train_pca, X_test_pca, y_train_pca, y_test_pca = train_test_split(X_pca, y, test_size=0.33, random_state=42)

print('Número reduzido de atributos:', X_pca.shape[1])

#----------------------------------------------------------------------------------------------------------

# Aplicando Train test split na seleção de atributos
X_train_sel, X_test_sel, y_train_sel, y_test_sel = train_test_split(X, y, test_size=0.33, random_state=42)


fvalue_selector = SelectKBest(f_classif, k=20)
X_kbest = fvalue_selector.fit_transform(X_train_sel, y_train_sel)
X_kbest2= fvalue_selector.fit_transform(X_test_sel, y_test_sel)
print('Número de atributos de Seleçao:', X_kbest.shape[1])


Número original de atributos: 64
Número reduzido de atributos: 21
Número de atributos de Seleçao: 20


/opt/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/_univariate_selection.py:114: UserWarning: Features [ 0 32 39] are constant.
  UserWarning)
/opt/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/_univariate_selection.py:115: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw
/opt/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/_univariate_selection.py:114: UserWarning: Features [ 0 16 24 32 39 56] are constant.
  UserWarning)
/opt/anaconda3/lib/python3.7/site-packages/sklearn/feature_selection/_univariate_selection.py:115: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


### Colocando tudo dentro de um dicionário e rodando com um loop

In [139]:
d = {'originais':[X_train, y_train, X_test, y_test],'redução':[X_train_pca, y_train_pca, X_test_pca, y_test_pca],
     'seleção':[X_kbest, y_train_sel, X_kbest2, y_test_sel]}

In [140]:
#essa parte é só um teste
# observação d[k][0] é X e d[k][1] é y
for k in d.keys():
    print(k)

originais
redução
seleção


In [141]:
#o d2 é mais um dicionário para armazenar todos os y_pred de cada modelo (ainda tenho que organiza-lo melhor)

#OBS: FALTA FAZER A FUNÇÃO DO MULTIPLE VOTES JOGAR O Y_PRED NA VARIÁVEL (AINDA NÃO CONSEGUI)

d2={}
for k in d.keys():
    print('==========================================================================')
    print('Tempos e acurácias dos modelos com os dados',k,':')
    print('==========================================================================')
    knn_bagging(d[k][0],d[k][1],d[k][2], d[k][3], k)  
    print('_____________________')
    decision_tree(d[k][0],d[k][1],d[k][2], d[k][3],k)
    #print('_____________________')
    multiples_vote(d[k][0],d[k][1],d[k][2], d[k][3])
    #print('_____________________')
    y_pred_g1 = gradient_boost(d[k][0],d[k][1],d[k][2], d[k][3],k)
    #d2[k + '_boost'] = y_pred_g1
    #print('\n')

Tempos e acurácias dos modelos com os dados originais :
Acurácia de KNeighbors simples: 0.9767150760719225
Tempo: 0.1627061367034912
Acurácia de KNeighbors Bagging (c/ 10 estimators): 0.9709163208852004
Tempo: 0.3684539794921875
Acurácia de KNeighbors Bagging (c/ 100 estimators): 0.9717461964038726
Tempo: 3.8232011795043945
-------------------------------------
Teste do McNemar
[[580, 10], [1, 3]]
statistic=1.000, p-value=0.012
Proporções de erros diferentes (rejeitou H0)
[[582, 8], [0, 4]]
statistic=0.000, p-value=0.008
Proporções de erros diferentes (rejeitou H0)
[[578, 3], [4, 9]]
statistic=3.000, p-value=1.000
Mesma proporção de erros (falhou em rejeitar H0)
_____________________
Acurácia de Decision Tree puro: 0.8204426002766253
Tempo: 0.06041073799133301
Acurácia de Random Forest: 0.9592842323651452
Tempo: 0.6413118839263916
Acurácia de Extreme Randomized Trees: 0.971753112033195
Tempo: 0.4829719066619873
-------------------------------------
Teste do McNemar
[[503, 3], [75, 13]]

In [146]:
analise = pd.DataFrame({'Modelo' : tipo,
                 'Acuracia' : acc,
                 'tempo' : time_l,
                 'tipo': dados })

In [147]:
analise.sort_values(by=['Acuracia'],ascending=False).head(10)

Modelo  Acuracia      tempo  \
0                     Acurácia de KNeighbors simples:  0.976715   0.162706   
9   Acurácia de KNeighbors Bagging (c/ 100 estimat...  0.973409   2.291666   
12                         Acurácia de Random Forest:  0.973396   0.537558   
5                          Acurácia de Random Forest:  0.971753   0.482972   
2   Acurácia de KNeighbors Bagging (c/ 100 estimat...  0.971746   3.823201   
1   Acurácia de KNeighbors Bagging (c/ 10 estimato...  0.970916   0.368454   
7                     Acurácia de KNeighbors simples:  0.969239   0.103337   
8   Acurácia de KNeighbors Bagging (c/ 10 estimato...  0.962593   0.253812   
4                          Acurácia de Random Forest:  0.959284   0.641312   
6                 Acurácia de Gradient Boosting Tree:  0.957611  11.849338   

         tipo  
0   originais  
9     redução  
12    redução  
5   originais  
2   originais  
1   originais  
7     redução  
8     redução  
4   originais  
6   originais

#### O modelo que teve a melhor acuracia com relacao ao tempo foi KNeighbors simples com os dados originais, sem tranformação ou seleção

Acurácia de KNeighbors simples: 0.9767150760719225 
Tempo: 0.24569296836853027

**** Para melhor visualização dos dados, anexamos o report em arquivo .docx